libraries/chain/controller.cpp

```diff
@@ -2694,6 +2701,16 @@ struct controller_impl {
 
    bool is_on_main_thread() { return main_thread_id == std::this_thread::get_id(); };
 
+   void set_to_write_window() {
+      app_window = app_window_type::write;
+   }
+   void set_to_read_window() {
+      app_window = app_window_type::read;
+   }
+   bool is_write_window() const {
+      return app_window == app_window_type::write;
+   }
```

Comment:
I think you should only have one of `set_db_read_only_mode` and `set_to_read_window`. Can you just use the db read only state? `context.control.db().is_read_only_mode()` Would need to add the `is_read_only_mode()` accessor to chainbase.

Reply:
We need to keep the both. `db_read_only_mode` is for protecting chainbase not to be updated while a read-only transaction is executed, either in read window or write window. It cannot tell whether the app() is in write window (tasks are running sequentially and safe to update global data like eos-vm-oc code cache) or in read window (tasks are running in parallel to each other).

Reply:
That makes sense. I'm not thinking of anyway to make this cleaner at the moment. I do think this is better than a registered function that calls `app().executor().is_write_window()`. 

You can change all the `producer_plugin` calls of `app().executor().is_write_window()` with `chain.is_write_window()`.